skyportal/facility_apis/ztf.py

```diff
@@ -146,6 +185,124 @@ def _build_observation_plan_payload(self, request):
+
+    Parameters
+    ----------
+    url : str
+        ZTF forced photometry service data file location.
+    altdata: dict
+        Contains ZTF photometry api_token for the user
+    df_request: pandas.DataFrame
+        DataFrame containing request parameters (ra, dec, start jd, end jd)
+    request_id : int
+        FollowupRequest SkyPortal ID
+    instrument_id : int
+        Instrument SkyPortal ID
+    """
+
+    from ..models import (
+        DBSession,
+        FollowupRequest,
+        Instrument,
+    )
+
+    Session = scoped_session(sessionmaker(bind=DBSession.session_factory.kw["bind"]))
+    session = Session()
+
+    try:
+        request = session.query(FollowupRequest).get(request_id)
+        instrument = session.query(Instrument).get(instrument_id)
+        r = requests.get(
+            url,
+            auth=HTTPBasicAuth(
+                altdata['ipac_http_user'], altdata['ipac_http_password']
+            ),
+        )
+        df = ascii.read(
+            r.content.decode(), header_start=0, data_start=1, comment='#'
+        ).to_pandas()
+
+        df.columns = df.columns.str.replace(',', '')
+        desired_columns = {
+            'jd',
+            'forcediffimflux',
+            'forcediffimfluxunc',
+            'diffmaglim',
+            'zpdiff',
+            'filter',
+        }
+        if not desired_columns.issubset(set(df.columns)):
+            raise ValueError('Missing expected column')
+        df['ra'] = df_request['ra']
+        df['dec'] = df_request['dec']
+        df.rename(
+            columns={'diffmaglim': 'limiting_mag'},
+            inplace=True,
+        )
+        df = df.replace({"null": np.nan})
+        df['mjd'] = astropy.time.Time(df['jd'], format='jd').mjd
+        df['filter'] = df['filter'].str.replace('_', '')
+        df['filter'] = df['filter'].str.lower()
+        df = df.astype({'forcediffimflux': 'float64', 'forcediffimfluxunc': 'float64'})
+
+        df['mag'] = df['zpdiff'] - 2.5 * np.log10(df['forcediffimflux'])
+        df['magerr'] = 1.0857 * df['forcediffimfluxunc'] / df['forcediffimflux']
+
+        snr = df['forcediffimflux'] / df['forcediffimfluxunc'] < 5
+        df['mag'].loc[snr] = None
+        df['magerr'].loc[snr] = None
+
+        iszero = df['forcediffimfluxunc'] == 0.0
+        df['mag'].loc[iszero] = None
+        df['magerr'].loc[iszero] = None
+
+        isnan = np.isnan(df['forcediffimflux'])
+        df['mag'].loc[isnan] = None
+        df['magerr'].loc[isnan] = None
+
+        df = df.replace({np.nan: None})
+
+        drop_columns = list(
+            set(df.columns.values)
+            - set(['mjd', 'ra', 'dec', 'mag', 'magerr', 'limiting_mag', 'filter'])
+        )
+
+        df.drop(
+            columns=drop_columns,
+            inplace=True,
+        )
+        df['magsys'] = 'ab'
+
+        data_out = {
+            'obj_id': request.obj_id,
+            'instrument_id': instrument.id,
+            'group_ids': 'all',
```

Comment:
Maybe groups associated with the current user?

Reply:
good idea!